src/Data/List/Relation/Unary/All/Properties.agda

```diff
@@ -505,6 +506,52 @@ module _ {P : A → Set p} {Q : A → Set q} (P? : Decidable P) where
   ... | false = filter⁺ Qxs
   ... | true  = Qx ∷ filter⁺ Qxs
 
+  filter⁻ : ∀ {xs} → All Q (filter P? xs) → All Q (filter (¬? ∘ P?) xs) → All Q xs
+  filter⁻ {[]}         []          []                           = []
+  filter⁻ {x ∷ xs}       all⁺        all⁻ with P? x  | ¬? (P? x)
+  filter⁻ {x ∷ xs}       all⁺        all⁻  | yes  Px | yes  ¬Px = contradiction Px ¬Px
+  filter⁻ {x ∷ xs} (qx ∷ all⁺)       all⁻  | yes  Px | no  ¬¬Px = qx ∷ filter⁻ all⁺ all⁻
+  filter⁻ {x ∷ xs}       all⁺  (qx ∷ all⁻) | no    _ | yes  ¬Px = qx ∷ filter⁻ all⁺ all⁻
+  filter⁻ {x ∷ xs}       all⁺        all⁻  | no  ¬Px | no  ¬¬Px = contradiction ¬Px ¬¬Px
+
+------------------------------------------------------------------------
+-- derun and deduplicate
+
+module _ {P : A → Set p} {Q : A → A → Set q} (Q? : B.Decidable Q) where
+
+  derun⁺ : ∀ {xs} → All P xs → All P (derun Q? xs)
+  derun⁺ {[]} [] = []
+  derun⁺ {x ∷ []} (px ∷ []) = px ∷ []
+  derun⁺ {x ∷ y ∷ xs} (px ∷ all[P,y∷xs]) with does (Q? x y)
+  ... | false = px ∷ derun⁺ all[P,y∷xs]
+  ... | true  = derun⁺ all[P,y∷xs]
+
+  deduplicate⁺ : ∀ {xs} → All P xs → All P (deduplicate Q? xs)
+  deduplicate⁺ [] = []
+  deduplicate⁺ {x ∷ _} (px ∷ all[P,xs]) = px ∷ filter⁺ (¬? ∘ Q? x) (deduplicate⁺ all[P,xs])
+
+  module _ (P-resp-Q : P B.Respects (flip Q)) where
+
+    private
+      derun⁻-aux : ∀ x xs → All P (derun Q? (x ∷ xs)) → All P (x ∷ xs)
```

Comment:
This would be neater if you moved it into a `where` block underneath `derun⁻`. You could then get rid of the second nested anonymous module block.